okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,245 @@
+                }
+                return null;
+            }
+
+            @Override
+            public Void visit(QosException.RetryOther exception) {
+                if (maxNumRelocations <= 0) {
+                    callback.onFailure(call,
+                            new IOException("Exceeded the maximum number of allowed "
+                                    + "redirects for initial URL: " + call.request().url()));
+                } else {
+                    Optional<HttpUrl> redirectTo = urls.redirectTo(request().url(),
+                            exception.getRedirectTo().toString());
+                    if (!redirectTo.isPresent()) {
+                        callback.onFailure(call,
+                                new IOException("Failed to determine valid redirect URL "
+                                        + "for '" + exception.getRedirectTo() + "' and base URLs "
+                                        + urls.getBaseUrls()));
+                    } else {
+                        Request redirectedRequest = request().newBuilder()
+                                .url(redirectTo.get())
+                                .build();
+                        client.newCallWithMutableState(
+                                redirectedRequest, backoffStrategy, maxNumRelocations - 1)
+                                .enqueue(callback);
+                    }
+                }
+                return null;
+            }
+
+            @Override
+            public Void visit(QosException.Unavailable exception) {
+                Optional<Duration> backoff = backoffStrategy.nextBackoff();
+                if (!backoff.isPresent()) {
+                    log.debug("Max number of retries exceeded, failing call");
+                    callback.onFailure(call,
+                            new IOException("Failed to complete the request due to a "
+                                    + "server-side QoS condition: 503", exception));
+                } else {
+                    log.debug("Rescheduling call after backoff",
+                            SafeArg.of("backoffMillis", backoff.get().toMillis()));
+                    scheduleExecution(callback, backoff.get());
+                }
+                return null;
+            }
+        };
+    }
+
+    @Override
+    public Response execute() throws IOException {
+        SettableFuture<Response> future = SettableFuture.create();
+        enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                future.setException(exception);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                future.set(response);
+            }
+        });
+
+        try {
+            return future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```

Comment:
Could we copy callum's test that verifies this works: https://github.com/palantir/http-remoting/pull/587/files#diff-39cfad0395ff8616ffa04724e0c47221R29

Reply:
Also why aren't we calling `getDelegate().cancel()`?

Reply:
calling cancel now, thanks

Reply:
added the test, but it's not clear to me what exactly that thing is testing?

Reply:
@uschi2000 Callum and I wrote that test because after trouble with our product on production, a jstack and metrics showed many more hanging threads than open connections.

The test times out on http-remoting develop, demonstrating that if an incoming dropwizard request handler thread is interrupted (because the client aborts), any in-flight http-remoting network calls ignore this interruption and stay open for (~10 minutes).

Reply:
OK. with this refactor, the test is a bit bogus: it passes even when the thread interrupt line is removed.